tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,416 @@
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk));
+                if (i == conn->psk_params.chosen_psk_wire_index) {
+                    EXPECT_OK(validate_psk_is_not_wiped(psk));
+                } else {
+                    EXPECT_OK(validate_psk_is_wiped(psk));
+                }
+            }
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out)); 
+        }
+    }
+
+    /* Functional test */
+    {
+        /* Setup connections */
+        EXPECT_SUCCESS(s2n_enable_tls13());
+        struct s2n_connection *client_conn, *server_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        struct s2n_config *config = NULL;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(client_conn, S2N_TLS13));
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(server_conn, S2N_TLS13));
+
+        struct s2n_psk *shared_psk = NULL;
+        uint16_t shared_psk_wire_index = 0;
+        EXPECT_OK(setup_order_mismatch_psks(server_conn, client_conn, &shared_psk, &shared_psk_wire_index));
+        EXPECT_NOT_NULL(shared_psk);
+
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&client_conn->handshake.io, &server_conn->handshake.io,
+                s2n_stuffer_data_available(&client_conn->handshake.io)));
+        EXPECT_SUCCESS(s2n_client_hello_recv(server_conn));
+
+        /* Verify shared PSK chosen */
+        EXPECT_EQUAL(server_conn->psk_params.chosen_psk_wire_index, shared_psk_wire_index);
+        EXPECT_EQUAL(server_conn->psk_params.chosen_psk->identity.size, sizeof(test_identity_1));
+        EXPECT_BYTEARRAY_EQUAL(server_conn->psk_params.chosen_psk->identity.data, test_identity_1, sizeof(test_identity_1));
+        EXPECT_EQUAL(server_conn->psk_params.chosen_psk->secret.size, sizeof(test_secret_1));
+        EXPECT_BYTEARRAY_EQUAL(server_conn->psk_params.chosen_psk->secret.data, test_secret_1, sizeof(test_secret_1));
+        EXPECT_EQUAL(server_conn->psk_params.chosen_psk->hmac_alg, shared_psk->hmac_alg);
```

Comment:
When you're writing tests, you can assume that the functionality you aren't testing works correctly (in this case, the client psk extension). You can still do sanity checks to validate you set things up right (like verifying that the ClientHello chose the PSK you expected to choose), but you don't have to verify that dependent functionality worked properly (like verifying that the PSK's fields were filled in right). Writing tests to check everything leads to confusing / difficult to maintain tests. Focus on what you NEED to test.